dotnet/samples/Concepts/ChatCompletion/OpenAI_ReasonedFunctionCalling.cs

```diff
@@ -0,0 +1,231 @@
+    [Fact]
+    public async Task QueryAssistantToExplainFunctionCallsBeforeExecutionAsync()
+    {
+        // Create a kernel with MistralAI chat completion and WeatherPlugin
+        Kernel kernel = CreateKernelWithPlugin<WeatherPlugin>();
+        kernel.AutoFunctionInvocationFilters.Add(new QueryExplainFunctionInvocationFilter(this.Output));
+        var service = kernel.GetRequiredService<IChatCompletionService>();
+
+        // Invoke chat prompt with auto invocation of functions enabled
+        var chatHistory = new ChatHistory
+        {
+            new ChatMessageContent(AuthorRole.User, "What is the weather like in Paris?")
+        };
+        var executionSettings = new OpenAIPromptExecutionSettings { ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions };
+        var result = await service.GetChatMessageContentAsync(chatHistory, executionSettings, kernel);
+        chatHistory.Add(result);
+        Console.WriteLine(result);
+    }
+
+    /// <summary>
+    /// This <see cref="IAutoFunctionInvocationFilter"/> will respond to function call requests and ask the model to explain why it is
+    /// calling the function(s). It is only suitable for transient use because it stores information about the functions that have been
+    /// called for a single chat history.
+    /// </summary>
+    private sealed class RespondExplainFunctionInvocationFilter : IAutoFunctionInvocationFilter
+    {
+        private readonly HashSet<string> _functionNames = [];
+
+        public async Task OnAutoFunctionInvocationAsync(AutoFunctionInvocationContext context, Func<AutoFunctionInvocationContext, Task> next)
+        {
+            // Get the function calls for which we need an explanation
+            var functionCalls = FunctionCallContent.GetFunctionCalls(context.ChatHistory.Last());
+            var needExplanation = 0;
+            foreach (var functionCall in functionCalls)
+            {
+                var functionName = $"{functionCall.PluginName}-{functionCall.FunctionName}";
+                if (_functionNames.Add(functionName))
+                {
+                    needExplanation++;
+                }
+            }
+
+            if (needExplanation > 0)
+            {
+                // Create a response asking why these functions are being called
+                context.Result = new FunctionResult(context.Result, $"Provide an explanation why you are calling function {string.Join(',', _functionNames)} and try again");
+                return;
+            }
+
+            // Invoke the functions
+            await next(context);
+        }
+    }
+
+    /// <summary>
+    /// This <see cref="IAutoFunctionInvocationFilter"/> uses the currently available <see cref="IChatCompletionService"/> to query the model
+    /// to find out what certain functions are being called.
+    /// </summary>
+    private sealed class QueryExplainFunctionInvocationFilter(ITestOutputHelper output) : IAutoFunctionInvocationFilter
```

Comment:
Should we also do an example where someone is using prompts? Since prompts also trigger function filters, we would need a check to make sure it doesn't try and invoke the LLM in that case, right?

Reply:
Added one new sample that uses a prompt